### PR TITLE
refactor: Infinitescroll Vue3 composition api

### LINKED ISSUE #4952: Infinitescroll Vue3 composition api
### spec

- use [VueUse](https://vueuse.org/core/useInfiniteScroll/#useinfinitescroll)
- rewrite `infiniteScrollMixin` Nuxt3 way (create composable)

follow up #4946

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring

## Context

- [x] Closes #4952
- [x] Refactor CollectionList component composition api
- [x] apply infinitescroll to both collection and items on explore page

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Review by roiLeo — COMMENTED** [review]
Here are my concern:
- `useDebounceFn` 😐 
- 200+ LOC
- DOM elements access (no ssr) 
- event listener

have you tried to use `vInfiniteScroll` from @vueuse/components?

**Comment by Jarsen136** [comment]
> Here are my concern:
> 
> * `useDebounceFn` 😐

It's necessary to use `debounce` if some events would be triggered frequently. In this scroll case, user would scroll to the bottom frequently to fetch next page.

> * DOM elements access (no ssr)

I'm really curious about how could we improve the ssr in this case. I could not thing of anything related to it TBH. Could you pls expand more?

> * event listener

We have to listen to the scroll event to know which page we are visiting.

> * 200+ LOC
> 
> have you tried to use `vInfiniteScroll` from @vueuse/components?

I have read the docs about [useInfiniteScroll](https://vueuse.org/core/useInfiniteScroll/). The function provided could not fit our requirements. It could only provide single-direction scrolling. In our codebase, the infinite scroll mixins provide much more functionality, like a double-direction scroll, prefetch data, pagination etc. That's why it has so much LOC :)

**Comment by roiLeo** [comment]
> It's necessary to use `debounce` if some events would be triggered frequently. In this scroll case, user would scroll to the bottom frequently to fetch next page.

hmmm, okay

> I'm really curious about how could we improve the ssr in this case. I could not think of anything related to it TBH. Could you pls expand more?

maybe use `ref` on tag

> We have to listen to the scroll event to know which page we are visiting.

`@scroll`  event?

> I have read the docs about [useInfiniteScroll](https://vueuse.org/core/useInfiniteScroll/). The function provided could not fit our requirements. It could only provide single-direction scrolling. In our codebase, the infinite scroll mixins provide much more functionality, like a double-direction scroll, prefetch data, pagination etc. That's why it has so much LOC :)

`direction?: "top" | "bottom" | "left" | "right"`


I'll take a look when I have time to make it work with vueuse. pretty sure it's doable

**Review by roiLeo — APPROVED** [review]

**Comment by vikiival** [comment]
Screenshot 2023-02-24 at 10 46 20


Up direction has up button anyway so I would check if user on first hit is on other page than one.
and for the direction going down I would prefer 
[useInfiniteScroll](https://vueuse.org/core/useInfiniteScroll/).


+ may https://vueuse.org/core/useOffsetPagination/#useoffsetpagination can help

**Comment by Jarsen136** [comment]
> 
> > I'm really curious about how could we improve the ssr in this case. I could not think of anything related to it TBH. Could you pls expand more?
> 
> maybe use `ref` on tag
Sorry, I just thought ssr is server-side render. I guess use ref does not change anything to it. Let me know if I miss something.

> 
> > We have to listen to the scroll event to know which page we are visiting.
> 
> `@scroll` event?
> 
However, it binds the scroll event on `window` for now.

> > I have read the docs about [useInfiniteScroll](https://vueuse.org/core/useInfiniteScroll/). The function provided could not fit our requirements. It could only provide single-direction scrolling. In our codebase, the infinite scroll mixins provide much more functionality, like a double-direction scroll, prefetch data, pagination etc. That's why it has so much LOC :)
> 
> `direction?: "top" | "bottom" | "left" | "right"`
> 
> I'll take a look when I have time to make it work with vueuse. pretty sure it's doable

emm. Maybe use twice time to bind two directions. I'm not sure if would work as expected. Anyway, I'm willing to see if someone could improve the code.

**Comment by Jarsen136** [comment]
> Up direction has up button anyway so I would check if user on first hit is on other page than one. 

The up button for loading the previous is only for the explore gallery page because it has a special structure. I wonder if we would like to apply the up button to other pages also. If so, we could do some refactoring and use [useInfiniteScroll](https://vueuse.org/core/useInfiniteScroll/). for the down-direction loading.

double-direction scroll example: https://beta.kodadot.xyz/rmrk/explore/collectibles?page=3

**Comment by yangwao** [comment]
hey haven't read anything on this pr, just having head's up on this that we should probably drop infinity scroll to the top and replace it to the button best which adds 4 rows with extra items

**Comment by Jarsen136** [comment]
> hey haven't read anything on this pr, just having head's up on this that we should probably drop infinity scroll to the top and replace it to the button best which adds 4 rows with extra items

OKi, I will make it work.

**Comment by Jarsen136** [comment]
I have created a `useListInfiniteScroll` composable function and applied it to the CollectionList component.

Here is the preview path:  https://deploy-preview-5108--koda-nuxt.netlify.app/rmrk/explore/collectibles

**Comment by yangwao** [comment]
> Here is the preview path

seems working well, can try to add to items?

**Comment by Jarsen136** [comment]
> > Here is the preview path
> 
> seems working well, can try to add to items?

Yes, I will refactor GalleryList component to composition api first, and then I would apply the infinitescroll to items.

**Comment by Jarsen136** [comment]
> > > Here is the preview path
> > 
> > 
> > seems working well, can try to add to items?
> 
> Yes, I will refactor GalleryList component to composition api first, and then I would apply the infinitescroll to items.

✅ added infiniteScroll to redesign items page

preview: https://deploy-preview-5108--koda-nuxt.netlify.app/bsx/explore/items?redesign=true

File: components/items/ItemsGrid/ItemsGrid.vue [hunk 1]
```diff
@@ -7,33 +7,74 @@
...
 
     <hr />
-    <DynamicGrid>
-      <div v-for="(nft, index) in nfts" :key="`${nft.id}=${index}`">
+    <a
+      v-if="startPage > 1 && !isLoading && total > 0"
+      class="is-flex is-justify-content-center pb-4"
+      @click="reachTopHandler">
+      <b-icon icon="chevron-up" />
+    </a>
```
Comment (vikiival) [inline components/items/ItemsGrid/ItemsGrid.vue:15]: just one thing, I would make this as component because we plan to reuse that across app

**Review by vikiival — APPROVED** [review]
Let it break 👀

File: components/items/ItemsGrid/ItemsGrid.vue (hunk 1 above)
Comment (Jarsen136) [reply components/items/ItemsGrid/ItemsGrid.vue]: DONE

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 34.13 USD/KSM ~ 1.465 $KSM
🧗 [Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG](https://kusama.subscan.io/account/Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)
🔗 [0xd2679d658fa1c638d1493d0d547b8cea83659620bd4d210502e9563a382fc732](https://kusama.subscan.io/extrinsic/0xd2679d658fa1c638d1493d0d547b8cea83659620bd4d210502e9563a382fc732)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**